Source/WebCore/page/Page.h

```diff
@@ -1390,8 +1390,31 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
     void updateDisplayEDRSuppression();
 #endif
 
-    // Checking hardware keyboard attached
-    void setHardwareKeyboardAttached(bool attached) { m_hardwareKeyboardAttached = attached; }
+    using HardwareKeyboardAttachmentObserver = Function<void(bool)>;
+    Vector<HardwareKeyboardAttachmentObserver> m_hardwareKeyboardAttachmentObservers;
+
+    void addHardwareKeyboardAttachmentObserver(HardwareKeyboardAttachmentObserver&& observer)
+    {
+        m_hardwareKeyboardAttachmentObservers.append(WTFMove(observer));
+    }
+
+    void flushHardwareKeyboardAttachmentObservers()
+    {
+        bool attached = m_hardwareKeyboardAttached;
+        std::ranges::for_each(m_hardwareKeyboardAttachmentObservers, [attached](auto& observer) {
+            observer(attached);
+        });
+        m_hardwareKeyboardAttachmentObservers.clear();
+    }
+
+    void didUpdateHardwareKeyboardAttachment(bool attached)
+    {
+        if (m_hardwareKeyboardAttached == attached)
+            return;
+
+        m_hardwareKeyboardAttached = attached;
+    }
+
     bool hardwareKeyboardAttached() const { return m_hardwareKeyboardAttached; }
```

Comment:
I understand you are storing this state internally in `WebCore::Page` to be able to carefully flush the observers, but no one should be able to read it. The ground truth should be read from `WebPage`, so I would suggest to remove this `hardwareKeyboardAttached()` getter from this class.